### ISSUE TITLE
Support restart on fail for executables and containers

### ISSUE THREAD

**Issue by bit365** [dotnet/aspire#1708]
When container A starts, the service in container A needs some startup time. At this time, service B has already started and is highly dependent on the service in container A. Therefore, service B ends due to an exception, such as the subscription dependency of Dapr sidecar. on the RabbitMQ service queue.

```csharp
builder.AddRabbitMQContainer("rabbitmq")
    .WithVolumeMount("./configs/rabbitmq/enabled_plugins.conf", "/etc/rabbitmq/enabled_plugins")
    .WithVolumeMount("./configs/rabbitmq/rabbitmq.conf", "/etc/rabbitmq/rabbitmq.conf")
    .WithEndpoint(1883, 1883)
    .WithEndpoint(15672, 15672)
    .WithEndpoint(5672, 5672)
    .WithEnvironment("TZ", "Asia/Shanghai");

var apiService = builder.AddProject<Projects.AspireApp_ApiService>("apiservice");
var apiServiceDapr = apiService.WithDaprSidecar(sidecarBuilder =>
{
    sidecarBuilder.WithOptions(new DaprSidecarOptions
    {
        AppId = "apiservice",
        ResourcesPaths = ["./components"],
        LogLevel = "debug"
    })
    .WithEnvironment("AMQP_CONNECTION_STRING", "amqp://test:hellotest@localhost:5672")
    .WithEnvironment("MQTT_CONNECTION_STRING", "tcp://test:hellotest@localhost:1883");
});
```
In the above code, the container has been started, but RabbitMQ in the container has not yet been started. At this time, the Dapr sidecar has been started and requires the RabbitMQ service. However, because the service cannot be found, the Dapr sidecar stops working and there is nothing to set. Try the restart strategy, and then although RabbitMQ has started successfully, Dapr becomes dead.

Wish there was an option to automatically try to restart after failure, so that would solve the problem.

The container can be restarted after failure through startup parameter settings, but I don't want to containerize the Dapr sidecar, so there is no option to restart.

```csharp
sidecarBuilder.WithRestartAlways()
.WithRestartAlways(10) // 10 seconds
```

Hopefully something like this setup.

**Comment by davidfowl** [comment]
@karolz-ms we discussed this a while ago. Since we don't have startup dependencies, we need a way to restart on fail when containers need to connect to each other, and we are unable to control retries inside of that container.

**Comment by karolz-ms** [comment]
Yep, we should be able to address it sometime before GA.

**Comment by rozbo** [comment]
+1

**Comment by karolz-ms** [comment]
> @karolz-ms we discussed this a while ago. Since we don't have startup dependencies, we need a way to restart on fail when containers need to connect to each other, and we are unable to control retries inside of that container.

We have also seen cases when a container starts successfully and fairly quickly, but then takes 10+ seconds to actually start responding to requests. So the startup dependencies would not help in that case.

Best solution is to make the client robust (retry connections with exponential backoff). Second best is to retry the failed containers/executables a few times before giving up, which is what this feature is about.

**Comment by bit365** [comment]
In fact, there are many ways to solve this problem, and you have to consider how to design it.

1. Try to restart continuously (number of times, exponential backoff algorithm, try time interval).
2. Use the health endpoint provided by the service in the container to monitor health status.
3. Start in sequence through certain dependency settings.

etc.

**Comment by maddymontaquila** [comment]
@bit365 - do you still want this feature now that we have .WaitFor() and the ability to stop/restart from the dashboard?

cc @karolz-ms @davidfowl

**Comment by koenbeuk** [comment]
> [@bit365](https://github.com/bit365) - do you still want this feature now that we have .WaitFor() and the ability to stop/restart from the dashboard?
> 
> cc [@karolz-ms](https://github.com/karolz-ms) [@davidfowl](https://github.com/davidfowl)

We do. In our case we have a long running project that may sporadically fail due to exceptions (e.g. network errors). We are looking for a feature that's similar to Dockers restart on-failure feature.

**Comment by davidfowl** [comment]
In local dev?

**Comment by koenbeuk** [comment]
Yes, as well as in our test environments. We have a long running process that may take hours or days to complete that we run both locally as well as in a dev environment. This process may fail at any time and can recover upon restart.

**Comment by KennethHoff** [comment]
I recently took over a legacy project using docker compose, and to make it easier for myself to migrate to .Net Aspire I started by simply calling `docker compose up <serviceName>` for all the services I've yet to migrate in the docker compose. This doesn't allow me to set env. variables or anything like that, but it does allow me to see the logs and start (not stop 😆 ) resources. This system is very flaky and at least one of the resources fails to start every time I start up Aspire, so being able to add say a `.WithRetries(5)` to all of these resouces would be great.

Yes, it would be a hack to get a hack running smoother, but if I encountered this situation I'm sure others would too.

**Comment by sunnamed434** [comment]
Would be amazing!

**Comment by karolz-ms** [comment]
It makes sense to do this in the app model and leverage Aspire health checks to determine the schedule of restart attempts

**Comment by joperezr** [comment]
As part of triage cleanup ahead of the 13.3 release, this issue is being moved to the Backlog milestone. If you believe it should remain in 13.3, feel free to re-add to the milestone and leave a comment here why it should not be punted.
